dpos/state/checkpoint.go

```diff
@@ -0,0 +1,527 @@
+package state
+
+import (
+	"io"
+
+	"github.com/elastos/Elastos.ELA/common"
+	"github.com/elastos/Elastos.ELA/core/types"
+)
+
+// KeyFrame holds necessary state about arbitrators
+type KeyFrame struct {
+	CurrentArbitrators [][]byte
+}
+
+// StateKeyFrame holds necessary state about State
+type StateKeyFrame struct {
+	NodeOwnerKeys     map[string]string // NodePublicKey as key, OwnerPublicKey as value
+	PendingProducers  map[string]*Producer
+	ActivityProducers map[string]*Producer
+	InactiveProducers map[string]*Producer
+	CanceledProducers map[string]*Producer
+	IllegalProducers  map[string]*Producer
+	Votes             map[string]*types.Output
+	Nicknames         map[string]struct{}
+	SpecialTxHashes   map[common.Uint256]struct{}
+	PreBlockArbiters  map[string]struct{}
+
+	EmergencyInactiveArbiters map[string]struct{}
+	VersionStartHeight        uint32
+	VersionEndHeight          uint32
+}
+
+// RewardData defines variables to calculate reward of a round
+type RewardData struct {
+	OwnerProgramHashes          []*common.Uint168
+	CandidateOwnerProgramHashes []*common.Uint168
+	OwnerVotesInRound           map[common.Uint168]common.Fixed64
+	TotalVotesInRound           common.Fixed64
+}
+
```

Comment:
Here can use a NewRewardData method instead of duplicate initializations.

Reply:
Fixed